File: binding/include/network/libqsslerror.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qsslerror.html
pub const qsslerror = struct {

    /// New constructs a new QSslError object.
    pub fn New() QtC.QSslError {
        return qtc.QSslError_new();
    }


    /// New2 constructs a new QSslError object.
    pub fn New2(errorVal: i32) QtC.QSslError {
        return qtc.QSslError_new2(@intCast(errorVal));
    }


    /// New3 constructs a new QSslError object.
    pub fn New3(errorVal: i32, certificate: ?*anyopaque) QtC.QSslError {
        return qtc.QSslError_new3(@intCast(errorVal), @ptrCast(certificate));
    }


    /// New4 constructs a new QSslError object.
    pub fn New4(other: ?*anyopaque) QtC.QSslError {
        return qtc.QSslError_new4(@ptrCast(other));
    }


    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QSslError_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QSslError_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QSslError_OperatorEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorNotEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QSslError_OperatorNotEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn Error(self: ?*anyopaque, ) i32 {
        return qtc.QSslError_Error(@ptrCast(self));
    }

    pub fn ErrorString(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSslError_ErrorString(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsslerror.ErrorString: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Certificate(self: ?*anyopaque, ) QtC.QSslCertificate {
        return qtc.QSslError_Certificate(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QSslError_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/sslerror.html#types
pub const enums = struct {
    pub const SslError = enum {
        pub const NoError: i32 = 0;
        pub const UnableToGetIssuerCertificate: i32 = 1;
        pub const UnableToDecryptCertificateSignature: i32 = 2;
        pub const UnableToDecodeIssuerPublicKey: i32 = 3;
        pub const CertificateSignatureFailed: i32 = 4;
        pub const CertificateNotYetValid: i32 = 5;
        pub const CertificateExpired: i32 = 6;
        pub const InvalidNotBeforeField: i32 = 7;
        pub const InvalidNotAfterField: i32 = 8;
        pub const SelfSignedCertificate: i32 = 9;
        pub const SelfSignedCertificateInChain: i32 = 10;
        pub const UnableToGetLocalIssuerCertificate: i32 = 11;
        pub const UnableToVerifyFirstCertificate: i32 = 12;
        pub const CertificateRevoked: i32 = 13;
        pub const InvalidCaCertificate: i32 = 14;
        pub const PathLengthExceeded: i32 = 15;
        pub const InvalidPurpose: i32 = 16;
        pub const CertificateUntrusted: i32 = 17;
        pub const CertificateRejected: i32 = 18;
        pub const SubjectIssuerMismatch: i32 = 19;
        pub const AuthorityIssuerSerialNumberMismatch: i32 = 20;
        pub const NoPeerCertificate: i32 = 21;
        pub const HostNameMismatch: i32 = 22;
        pub const NoSslSupport: i32 = 23;
        pub const CertificateBlacklisted: i32 = 24;
        pub const CertificateStatusUnknown: i32 = 25;
        pub const OcspNoResponseFound: i32 = 26;
        pub const OcspMalformedRequest: i32 = 27;
        pub const OcspMalformedResponse: i32 = 28;
        pub const OcspInternalError: i32 = 29;
        pub const OcspTryLater: i32 = 30;
        pub const OcspSigRequred: i32 = 31;
        pub const OcspUnauthorized: i32 = 32;
        pub const OcspResponseCannotBeTrusted: i32 = 33;
        pub const OcspResponseCertIdUnknown: i32 = 34;
        pub const OcspResponseExpired: i32 = 35;
        pub const OcspStatusUnknown: i32 = 36;
        pub const UnspecifiedError: i32 = -1;
    };

};

File: binding/src/network/libqsslerror.cpp
#include <QDebug>
#include <QSslCertificate>
#include <QSslError>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qsslerror.h>
#include "libqsslerror.h"
#include "libqsslerror.hxx"

QSslError* QSslError_new() {
	 return new QSslError();
}

QSslError* QSslError_new2(int errorVal) {
	 return new QSslError(static_cast<QSslError::SslError>(errorVal));
}

QSslError* QSslError_new3(int errorVal, const QSslCertificate* certificate) {
	 return new QSslError(static_cast<QSslError::SslError>(errorVal), *certificate);
}

QSslError* QSslError_new4(const QSslError* other) {
	 return new QSslError(*other);
}

void QSslError_Swap(QSslError* self, QSslError* other) {
	self->swap(*other);
}

void QSslError_OperatorAssign(QSslError* self, const QSslError* other) {
	self->operator=(*other);
}

bool QSslError_OperatorEqual(const QSslError* self, const QSslError* other) {
	return self->operator==(*other);
}

bool QSslError_OperatorNotEqual(const QSslError* self, const QSslError* other) {
	return self->operator!=(*other);
}

int QSslError_Error(const QSslError* self) {
	return self->error();
}

libqt_string QSslError_ErrorString(const QSslError* self) {
	QString _ret = self->errorString();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QSslCertificate* QSslError_Certificate(const QSslError* self) {
	return new QSslCertificate(self->certificate());
}

void QSslError_Delete(QSslError* self) {
    delete self;
}

